File: models2.R
# =============================================================
# models2.R
# R-Uebersetzung von models2.py by Claude AI
# =============================================================

library(readxl)      # Excel einlesen
library(dplyr)        # Datenmanipulation
library(caret)         # createDataPartition (stratifizierter Split), Metriken
library(randomForest)   # Random Forest
library(rpart)            # Decision Tree
library(rpart.plot)       # Decision Tree Plot
library(xgboost)           # XGBoost
library(pROC)               # ROC / AUC
library(ggplot2)             # Plots

dir.create("pictures", showWarnings = FALSE)

# -------------------------------------------------------------
# Hilfsfunktionen
# -------------------------------------------------------------

conf_matrix <- function(y_test, y_pred, model) {
  cm <- table(factor(y_test, levels = c(0, 1)), factor(y_pred, levels = c(0, 1)))
  tn <- cm[1, 1]
  fp <- cm[1, 2]
  fn <- cm[2, 1]
  tp <- cm[2, 2]

  accuracy  <- (tp + tn) / (tp + tn + fp + fn)
  precision <- tp / (tp + fp)
  recall    <- tp / (tp + fn)
  f1        <- 2 * (precision * recall) / (precision + recall)

  cat(sprintf("------%s---------\n", model))
  print(cm)
  cat(sprintf("Accuracy: %.2f%%\n", accuracy * 100))
  cat(sprintf("Precision: %.2f%%\n", precision * 100))
  cat(sprintf("Recall (TPR): %.2f%%\n", recall * 100))
  cat(sprintf("F1 Score: %.2f%%\n", f1 * 100))
  cat("---------------\n")

  list(accuracy = accuracy, precision = precision, recall = recall, f1 = f1)
}

threshold_analysis <- function(y_test, y_prob, model_name = "Model") {
  # Sucht den besten Threshold basierend auf F1-Score

  thresholds <- seq(0.01, 0.9, length.out = 100)
  accuracies <- numeric(length(thresholds))
  precisions <- numeric(length(thresholds))
  recalls    <- numeric(length(thresholds))
  f1_scores  <- numeric(length(thresholds))

  for (i in seq_along(thresholds)) {
    t <- thresholds[i]
    y_pred <- as.integer(y_prob >= t)

    tp <- sum(y_pred == 1 & y_test == 1)
    tn <- sum(y_pred == 0 & y_test == 0)
    fp <- sum(y_pred == 1 & y_test == 0)
    fn <- sum(y_pred == 0 & y_test == 1)

    accuracies[i] <- (tp + tn) / (tp + tn + fp + fn)
    precisions[i] <- if ((tp + fp) == 0) 0 else tp / (tp + fp)
    recalls[i]    <- if ((tp + fn) == 0) 0 else tp / (tp + fn)
    f1_scores[i]  <- if ((precisions[i] + recalls[i]) == 0) 0 else
      2 * (precisions[i] * recalls[i]) / (precisions[i] + recalls[i])
  }

  plot_df <- data.frame(
    threshold = rep(thresholds, 4),
    value = c(accuracies, precisions, recalls, f1_scores),
    metric = rep(c("Accuracy", "Precision", "Recall", "F1 Score"), each = length(thresholds))
  )

  p <- ggplot(plot_df, aes(x = threshold, y = value, color = metric)) +
    geom_line() +
    labs(x = "Threshold", y = "a.u", color = NULL) +
    theme_minimal(base_size = 14)

  ggsave(filename = sprintf("pictures/thresholds_%s.pdf", model_name),
         plot = p, width = 8, height = 5, bg = "white")
  print(p)

  best_index <- which.max(f1_scores)
  cat("Best threshold based on F1:\n")
  cat(sprintf("Threshold: %.3f\n", thresholds[best_index]))
  cat(sprintf("F1 Score: %.3f\n", f1_scores[best_index]))
  cat(sprintf("Precision: %.3f\n", precisions[best_index]))
  cat(sprintf("Recall: %.3f\n", recalls[best_index]))
  cat(sprintf("Accuracy: %.3f\n", accuracies[best_index]))
}

benchmark <- function(roc_obj, model) {
  cat(sprintf("------%s---------\n", model))
  auc_val <- as.numeric(auc(roc_obj))
  cat(sprintf("AUC: %.2f\n", auc_val))
  gini <- 2 * auc_val - 1
  cat(sprintf("Gini: %.2f\n", gini))
  cat("---------------\n")
}

# -------------------------------------------------------------
# Daten einlesen
# -------------------------------------------------------------

df <- read_excel("data/Assignment_Dataset_2.xlsx", sheet = 1)
features <- c("DSRI", "GMI", "AQI", "SGI", "DEPI", "SGAI", "ACCR", "LEVI")

X <- as.matrix(df[, features])
Y <- ifelse(df$Manipulater == "Yes", 1, 0)

# -------------------------------------------------------------
# Train/Test Split (stratifiziert, analog train_test_split mit stratify=Y)
# -------------------------------------------------------------

set.seed(123)  # gleiche Samples im ganzen Skript
train_index <- createDataPartition(Y, p = 0.8, list = FALSE)

X_train <- X[train_index, ]
X_test  <- X[-train_index, ]
y_train <- Y[train_index]
y_test  <- Y[-train_index]

# -------------------------------------------------------------
# Skalierung
# -------------------------------------------------------------

scaler_center <- apply(X_train, 2, mean)
scaler_scale  <- apply(X_train, 2, sd)

X_train_scaled <- scale(X_train, center = scaler_center, scale = scaler_scale)
X_test_scaled  <- scale(X_test, center = scaler_center, scale = scaler_scale)

# -------------------------------------------------------------
# Logistische Regression (class_weight="balanced" -> inverse Gewichte)
# -------------------------------------------------------------

class_freq <- table(y_train)
weights_log <- ifelse(y_train == 1,
                       unname(length(y_train) / (2 * class_freq["1"])),
                       unname(length(y_train) / (2 * class_freq["0"])))

log_data <- data.frame(X_train_scaled)
log_data$y <- y_train

log_model <- glm(y ~ ., data = log_data, family = binomial(),
                  weights = weights_log)

test_data_log <- data.frame(X_test_scaled)
y_prob_log <- predict(log_model, newdata = test_data_log, type = "response")
threshold_log <- 0.5
y_pred_log <- as.integer(y_prob_log >= threshold_log)
log_res <- conf_matrix(y_test, y_pred_log, "Logistic Regression")

# -------------------------------------------------------------
threshold_analysis(y_test, y_prob_log, "Logistic Regression")

# -------------------------------------------------------------
# Random Forest
# -------------------------------------------------------------
# GridSearchCV wurde bereits offline durchgefuehrt, bestes Ergebnis:
# max_depth=5 (~ als maxnodes/Tiefenbegrenzung in R nicht 1:1 uebertragbar,
# daher ueber nodesize/maxnodes angenaehert), min_samples_leaf=5,
# min_samples_split=2, n_estimators=500

rf_data <- data.frame(X_train)
rf_data$y <- factor(y_train, levels = c(0, 1))

# class_weight="balanced" -> classwt umgekehrt proportional zu Klassenhaeufigkeit
# unname() ist noetig, da class_freq["0"]/["1"] selbst schon Namen tragen und
# c("0" = class_freq["0"], ...) sonst verschachtelte Namen wie "0.0" erzeugt,
# die randomForest nicht als gueltige Klassenlabel erkennt
classwt <- c("0" = unname(length(y_train) / (2 * class_freq["0"])),
             "1" = unname(length(y_train) / (2 * class_freq["1"])))

set.seed(123)
rf_model <- randomForest(
  y ~ .,
  data = rf_data,
  ntree = 500,
  nodesize = 5,       # ~ min_samples_leaf
  classwt = classwt,
  maxnodes = 2^5       # ~ Begrenzung analog max_depth=5
)

y_prob_rf <- predict(rf_model, newdata = data.frame(X_test), type = "prob")[, "1"]
threshold_rf <- 0.5
y_pred_rf <- as.integer(y_prob_rf >= threshold_rf)
rf_res <- conf_matrix(y_test, y_pred_rf, "Random Forest")

# -------------------------------------------------------------
threshold_analysis(y_test, y_prob_rf, "Random Forest")

# -------------------------------------------------------------
# RF Feature Importance
# -------------------------------------------------------------

importance_vals <- importance(rf_model)[, "MeanDecreaseGini"]
importance_df <- data.frame(feature = names(importance_vals), importance = importance_vals)
importance_df <- importance_df[order(-importance_df$importance), ]

cat("Importance Random forest\n")
print(importance_df)

p_imp <- ggplot(importance_df, aes(x = reorder(feature, -importance), y = importance)) +
  geom_col() +
  labs(x = NULL, y = "Importance") +
  theme_minimal(base_size = 14) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

ggsave("pictures/rf_feature_importance2.pdf", plot = p_imp, bg = "white")
print(p_imp)

# -------------------------------------------------------------
# Decision Tree
# -------------------------------------------------------------

dt_data <- data.frame(X_train)
dt_data$y <- factor(y_train, levels = c(0, 1))

# class_weight="balanced" ueber Prior-Wahrscheinlichkeiten angenaehert
prior_0 <- 0.5
prior_1 <- 0.5

dt_model <- rpart(
  y ~ .,
  data = dt_data,
  method = "class",
  parms = list(prior = c(prior_0, prior_1)),
  control = rpart.control(maxdepth = 5, minsplit = 2, minbucket = 5)
)

y_prob_dt <- predict(dt_model, newdata = data.frame(X_test), type = "prob")[, "1"]
threshold_dt <- 0.5
y_pred_dt <- as.integer(y_prob_dt >= threshold_dt)
dt_res <- conf_matrix(y_test, y_pred_dt, "Decision Tree")

# -------------------------------------------------------------
threshold_analysis(y_test, y_prob_dt, "Decision Tree")

# -------------------------------------------------------------
pdf("pictures/decision_tree.pdf")
rpart.plot(dt_model, extra = 104, fallen.leaves = TRUE, box.palette = "auto")
dev.off()

# -------------------------------------------------------------
# XGBoost
# -------------------------------------------------------------
# GridSearchCV wurde bereits offline durchgefuehrt, bestes Ergebnis:
# colsample_bytree=0.8, learning_rate=0.1, max_depth=4,
# min_child_weight=5, n_estimators=600, subsample=0.8
# Bester F1: 0.4488311688311688

dtrain <- xgb.DMatrix(data = X_train, label = y_train)
dtest  <- xgb.DMatrix(data = X_test, label = y_test)

xgb_params <- list(
  objective = "binary:logistic",
  eval_metric = "logloss",
  max_depth = 4,
  eta = 0.1,               # entspricht learning_rate
  min_child_weight = 5,
  subsample = 0.8,
  colsample_bytree = 0.8,
  scale_pos_weight = 1200 / 39
)

set.seed(123)
xgb_model <- xgb.train(
  params = xgb_params,
  data = dtrain,
  nrounds = 600           # entspricht n_estimators
)

y_prob_xgb <- predict(xgb_model, dtest)
threshold_xgb <- 0.5
y_pred_xgb <- as.integer(y_prob_xgb >= threshold_xgb)
xgb_res <- conf_matrix(y_test, y_pred_xgb, "XGBoost")

# -------------------------------------------------------------
threshold_analysis(y_test, y_prob_xgb, "XGBoost")

# -------------------------------------------------------------
# ROC Kurven
# -------------------------------------------------------------

roc_log <- roc(y_test, y_prob_log, quiet = TRUE)
roc_rf  <- roc(y_test, y_prob_rf, quiet = TRUE)
roc_dt  <- roc(y_test, y_prob_dt, quiet = TRUE)
roc_xgb <- roc(y_test, y_prob_xgb, quiet = TRUE)

roc_df <- rbind(
  data.frame(fpr = 1 - roc_log$specificities, tpr = roc_log$sensitivities, model = "Logistic Regression"),
  data.frame(fpr = 1 - roc_rf$specificities, tpr = roc_rf$sensitivities, model = "Random Forest"),
  data.frame(fpr = 1 - roc_dt$specificities, tpr = roc_dt$sensitivities, model = "Decision Tree"),
  data.frame(fpr = 1 - roc_xgb$specificities, tpr = roc_xgb$sensitivities, model = "XGBoost")
)

p_roc <- ggplot(roc_df, aes(x = fpr, y = tpr, color = model)) +
  geom_line() +
  geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "gray") +
  labs(x = "False Positive Rate", y = "True Positive Rate", color = NULL) +
  coord_equal() +
  theme_minimal(base_size = 14)

ggsave("pictures/roc_2.pdf", plot = p_roc, width = 6, height = 6, bg = "white")
print(p_roc)

benchmark(roc_log, "Logistic Regression")
benchmark(roc_dt, "Decision Tree")
benchmark(roc_rf, "Random Forest")
benchmark(roc_xgb, "XGBoost")

File: models2.py
# %%
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from latex_io import format_matrix
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier, plot_tree
from sklearn.metrics import confusion_matrix, roc_curve, auc
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from xgboost import XGBClassifier

plt.rcParams.update(
    {
        "font.size": 14,
        "axes.titlesize": 16,
        "axes.labelsize": 14,
        "xtick.labelsize": 14,
        "ytick.labelsize": 14,
        "legend.fontsize": 14,
        "figure.titlesize": 18,
    }
)
plt.rcParams["figure.figsize"] = (8, 5)


# %%
def conf_matrix(y_test, y_pred, model):
    cm = confusion_matrix(y_test, y_pred)
    tn = cm[0, 0]
    fp = cm[0, 1]
    fn = cm[1, 0]
    tp = cm[1, 1]
    accuracy = (tp + tn) / (tp + tn + fp + fn)
    precision = tp / (tp + fp)
    recall = tp / (tp + fn)
    f1 = 2 * (precision * recall) / (precision + recall)
    print(f"------{model}---------")
    print(cm)
    print(f"Accuracy: {accuracy*100:.2f}%")
    print(f"Precision: {precision*100:.2f}%")
    print(f"Recall (TPR): {recall*100:.2f}%")
    print(f"F1 Score: {f1*100:.2f}%")
    print("---------------")
    return accuracy, precision, recall, f1


def threshold_analysis(y_test, y_prob, model_name="Model"):
    """
    find best threshold

    Args:
        y_test (_type_): _description_
        y_prob (_type_): _description_
        model_name (str, optional): _description_. Defaults to "Model".
    """

    thresholds = np.linspace(0.01, 0.9, 100)
    accuracies = []
    precisions = []
    recalls = []
    f1_scores = []
    for t in thresholds:
        y_pred = (y_prob >= t).astype(int)
        accuracies.append(
            accuracy_score(y_test, y_pred)
        )  # hehe of course its a build in function in metrics
        precisions.append(precision_score(y_test, y_pred, zero_division=0))
        recalls.append(recall_score(y_test, y_pred, zero_division=0))
        f1_scores.append(f1_score(y_test, y_pred, zero_division=0))

    # fig, axes = plt.subplots(2, 2, figsize=(12, 8))
    metrics = [
        ("Accuracy", accuracies),
        ("Precision", precisions),
        ("Recall", recalls),
        ("F1 Score", f1_scores),
    ]
    for name, values in metrics:
        plt.plot(thresholds, values, label=name)
        plt.grid()
        plt.ylabel("a.u")
        plt.xlabel("Threshold")
    plt.legend()
    # for ax, (name, values) in zip(axes.flatten(), metrics):
    #    axes[0,0].plot(thresholds, values, label = name)
    #    #ax.set_title(f"{model_name}: {name}")
    #    ax.set_xlabel("Threshold")
    #    ax.set_ylabel(name)
    #    ax.grid(True)
    #    ax.legend()
    plt.tight_layout()
    plt.savefig(f"pictures/thresholds_{model_name}.pdf", bbox_inches="tight")
    plt.show()
    best_index = np.argmax(f1_scores)
    print(f"Best threshold based on F1:")
    print(f"Threshold: {thresholds[best_index]:.3f}")
    print(f"F1 Score: {f1_scores[best_index]:.3f}")
    print(f"Precision: {precisions[best_index]:.3f}")
    print(f"Recall: {recalls[best_index]:.3f}")
    print(f"Accuracy: {accuracies[best_index]:.3f}")


# %%
# import dataset
df = pd.read_excel("data/Assignment_Dataset_2.xlsx", sheet_name=0)
features = ["DSRI", "GMI", "AQI", "SGI", "DEPI", "SGAI", "ACCR", "LEVI"]
X = df[features].to_numpy()
Y = df["Manipulater"].map({"No": 0, "Yes": 1}).values

# %%
X_train, X_test, y_train, y_test = train_test_split(
    X,
    Y,
    test_size=0.2,
    random_state=123,  # keep same samples in whole script
    stratify=Y,  # keep ratio from manipulater in X,Y
)

# %%
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# %%
# Logistic Regression model
log_model = LogisticRegression(
    random_state=123,
    max_iter=1000,
    class_weight="balanced",
    penalty="l2",
    solver="liblinear",
)
log_model.fit(X_train_scaled, y_train)
y_prob_log = log_model.predict_proba(X_test_scaled)[:, 1]
threshold_log = 0.5
y_pred_log = (y_prob_log >= threshold_log).astype(int)
log_res = conf_matrix(y_test, y_pred_log, "Logistic Regression")

# %%
threshold_analysis(y_test, y_prob_log, "Logistic Regression")

# %%
## Random Forest model
# param_grid = {
#    "n_estimators": [200, 500],
#    "max_depth": [5, 8, 12, None],
#    "min_samples_leaf": [1, 3, 5],
#    "min_samples_split": [2, 5, 10],
# }
# rf = RandomForestClassifier(random_state=123, class_weight="balanced_subsample")
# grid = GridSearchCV(rf, param_grid, scoring="f1", cv=5, n_jobs=-1)
# grid.fit(X_train, y_train)
# print(grid.best_params_)
## result
## {'max_depth': 5, 'min_samples_leaf': 5, 'min_samples_split': 2, 'n_estimators': 500}

# %%
# Contiue Random Forest
rf_model = RandomForestClassifier(
    class_weight="balanced",
    random_state=123,
    max_depth=5,
    min_samples_leaf=5,
    min_samples_split=2,
    n_estimators=500,
)
rf_model.fit(X_train, y_train)  # not scaled
y_prob_rf = rf_model.predict_proba(X_test)[:, 1]
threshold_rf = 0.5
y_pred_rf = (y_prob_rf >= threshold_rf).astype(int)
# for t_rf in [0.5, 0.4, 0.3, 0.2, 0.1]:
#    y_pred_rf = (y_prob_rf >= t_rf).astype(int)
#    print(f"threshold: {t_rf}")
#    rf_res = conf_matrix(y_test, y_pred_rf, "Random Forest")
#    accuracy, precision, recall, f1 = rf_res
#    print("\n")
rf_res = conf_matrix(y_test, y_pred_rf, "Random Forest")

# %%
threshold_analysis(y_test, y_prob_rf, "Random Forest")

# %%
# RF feature importance
# additional random forest analysis
importance = pd.Series(rf_model.feature_importances_, index=features).sort_values(
    ascending=False
)
print("Importance Random forest")
print(importance)
importance.plot(kind="bar")
plt.ylabel("Importance")
plt.savefig(f"pictures/rf_feature_importance2.pdf", bbox_inches="tight")
plt.show()

# %%
# Decision Tree model
dt_model = DecisionTreeClassifier(
    random_state=123,
    class_weight="balanced",
    max_depth=5,
    min_samples_leaf=5,
    min_samples_split=2,
)
dt_model.fit(X_train, y_train)

y_prob_dt = dt_model.predict_proba(X_test)[:, 1]
threshold_dt = 0.5
y_pred_dt = (y_prob_dt >= threshold_dt).astype(int)

dt_res = conf_matrix(y_test, y_pred_dt, "Decision Tree")

# %%
threshold_analysis(y_test, y_prob_dt, "Decision Tree")

# %%
plot_tree(
    dt_model,
    feature_names=features,
    class_names=["No", "Yes"],
    filled=True,
    rounded=True,
)
plt.savefig("pictures/decision_tree.pdf", dpi=300, bbox_inches="tight")

plt.close()


# %%
# xgb_model = XGBClassifier(
#    random_state=123,
#    objective="binary:logistic",
#    eval_metric="logloss",
#    scale_pos_weight=1200 / 39,
# )
#
# param_grid = {
#    "max_depth": [3, 4, 5],
#    "learning_rate": [0.01, 0.05, 0.1],
#    "n_estimators": [200, 400, 600],
#    "min_child_weight": [1, 3, 5],
#    "subsample": [0.8, 1.0],
#    "colsample_bytree": [0.8, 1.0],
# }
#
# grid = GridSearchCV(
#    estimator=xgb_model, param_grid=param_grid, scoring="f1", cv=5, n_jobs=-1
# )
#
## Training!
# grid.fit(X_train, y_train)
#
# print("Beste Parameter:")
# print(grid.best_params_)
#
# print("Bester F1:")
# print(grid.best_score_)
## result
## Beste Parameter:
## {'colsample_bytree': 0.8, 'learning_rate': 0.1, 'max_depth': 4, 'min_child_weight': 5, 'n_estimators': 600, 'subsample': 0.8}
## Bester F1:
## 0.4488311688311688

# %%
# XGBoost model
params = {
    "colsample_bytree": 0.8,
    "learning_rate": 0.1,
    "max_depth": 4,
    "min_child_weight": 5,
    "n_estimators": 600,
    "subsample": 0.8,
}
xgb_model = XGBClassifier(
    random_state=123,
    objective="binary:logistic",
    eval_metric="logloss",
    scale_pos_weight=1200 / 39,
    **params,
)

xgb_model.fit(X_train, y_train)

y_prob_xgb = xgb_model.predict_proba(X_test)[:, 1]
threshold_xgb = 0.5
y_pred_xgb = (y_prob_xgb >= threshold_xgb).astype(int)
xgb_res = conf_matrix(y_test, y_pred_xgb, "XGBoost")
# for t_xgb in [0.5, 0.4, 0.3, 0.2, 0.1]:
#    y_pred_xgb = (y_prob_xgb >= t_xgb).astype(int)
#    print(f"Threshold: {t_xgb}")
#    res = conf_matrix(y_test, y_pred_xgb, "XGBoost")
#    print("\n")

# %%
threshold_analysis(y_test, y_prob_xgb, "XGBoost")

# %%
# roc curves
fpr_log, tpr_log, thresholds_log = roc_curve(y_test, y_prob_log)
fpr_rf, tpr_rf, thresholds_rf = roc_curve(y_test, y_prob_rf)
fpr_dt, tpr_dt, _ = roc_curve(y_test, y_prob_dt)
fpr_xgb, tpr_xgb, _ = roc_curve(y_test, y_prob_xgb)
plt.figure(figsize=(6, 6))
plt.plot(fpr_log, tpr_log, label="Logistic Regression")
plt.plot(fpr_rf, tpr_rf, label="Random Forest")
plt.plot(fpr_dt, tpr_dt, label="Decision Tree")
plt.plot(fpr_xgb, tpr_xgb, label="XGBoost")
plt.plot([0, 1], [0, 1], "--", color="gray", label="Random classifier")
plt.xlabel("False Positive Rate")
plt.ylabel("True Positive Rate")
plt.legend()
plt.grid()
plt.tight_layout()
plt.savefig(f"pictures/roc_2.pdf", bbox_inches="tight")
plt.show()


def benchmark(fpr, tpr, model):
    print(f"------{model}---------")
    auc_val = auc(fpr, tpr)
    print(f"AUC: {auc_val:.2f}")
    gini = 2 * auc_val - 1
    print(f"Gini: {gini:.2f}")
    print("---------------")


benchmark(fpr_log, tpr_log, "Logistic Regression")
benchmark(fpr_dt, tpr_dt, "Decision Tree")
benchmark(fpr_rf, tpr_rf, "Random Forest")
benchmark(fpr_xgb, tpr_xgb, "XGBoost")


# %%
